### PR TITLE
WIP, first pass at abstracting out TF functionality into a TFAlphaZeroProblem class

### PR DESCRIPTION
**Pull request by pstjohn**
Opening a PR w/ dave's changes for comments

### CODE REVIEW

File: rlmolecule/alphazero/tfalphazero_problem.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                status = self.policy_model.load_weights(self._checkpoint)
+                status.assert_existing_objects_matched()
+                logger.info(f'Loaded checkpoint {self._checkpoint}')
+            elif new_checkpoint == self._checkpoint:
+                logger.info(f'Skipping already loaded {self._checkpoint}')
+            else:
+                logger.info('No checkpoint found')
+
+    @abstractmethod
+    def _get_network_inputs(self, state: GraphSearchState) -> Dict:
```
Comment (pstjohn) [inline rlmolecule/alphazero/tfalphazero_problem.py:146]: we probably don't want the leading underscore if this is a public method implemented by the user. Is 'get_network_inputs` intuitive? Or maybe `get_policy_inputs`?

File: rlmolecule/alphazero/tfalphazero_problem.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+                 input_masks: dict = {},  # will default to 0 if you don't specify
+                 **kwargs):
+
+        super().__init__(**kwargs)
+        self._policy_model = policy_model
+
+        # Here we create default 0-values for masks, and then update with any 
+        # user supplied masks keyed on the input layer names.
+        self._input_masks = {inp.name: 0. for inp in self._policy_model.inputs}
+        self._input_masks.update(input_masks)
```
Comment (pstjohn) [inline rlmolecule/alphazero/tfalphazero_problem.py:67]: this looks great -- do we need to worry about dtypes and whether these are floats, numpy, or tf.constants?

**Review by pstjohn — COMMENTED** [review]
looking great! Thanks!

File: rlmolecule/alphazero/tfalphazero_problem.py (hunk 1 above)
Comment (davebiagioni) [reply rlmolecule/alphazero/tfalphazero_problem.py:146]: yeah, agreed.  we use policy most places (rather than "network") so that would be better.

File: rlmolecule/alphazero/tfalphazero_problem.py (hunk 2 above)
Comment (davebiagioni) [reply rlmolecule/alphazero/tfalphazero_problem.py:67]: hmm, the user should definitely be able to supply just a float. but we should definitely cast it correctly.  the default dtype for everything we do is tf.float32, yes?

Comment (davebiagioni) [reply rlmolecule/alphazero/tfalphazero_problem.py]: nm, I've been mentally stuck in continuous input land for awhile.  added a small helper function to both set default mask values and cast them according to layer type.

File: rlmolecule/alphazero/tfalphazero_problem.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        policy_inputs = [self.get_policy_inputs(parent)
+                         for state in itertools.chain((parent,), parent.get_next_actions())]
+
+        policy_inputs = {key: pad_sequences(
+                                [elem[key] for elem in policy_inputs], 
+                                padding='post',
+                                value=self.mask_dict[key])
+                         for key in policy_inputs[0]}
+
+        return policy_inputs['position'], policy_inputs['steps']
```
Comment (pstjohn) [inline rlmolecule/alphazero/tfalphazero_problem.py:198]: this is is probably still hallway-specific

File: rlmolecule/alphazero/tfalphazero_problem.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        # Update child nodes with predicted prior_logits
+        children_priors = {vertex: prior for vertex, prior in zip(parent.children, priors)}
+        value = float(tf.nn.sigmoid(values[0]))
+
+        return value, children_priors
+
+
+    def _get_policy_inputs_from_serialized_parent(
+            self,
+            serialized_parent: tf.Tensor) -> Tuple[dict, dict]:
```
Comment (pstjohn) [inline rlmolecule/alphazero/tfalphazero_problem.py:185]: This probably just a list of numpy arrays

**Comment by pstjohn** [comment]
sorry for the messy commit history, I had made a change in ORM refactor and wanted to bring that in
